python/ray/serve/tests/test_cluster.py

```diff
@@ -0,0 +1,110 @@
+import os
+import sys
+import time
+
+import pytest
+import requests
+
+import ray
+from ray import serve
+from ray.cluster_utils import Cluster
+
+
+@pytest.fixture
+def ray_cluster():
+    cluster = Cluster()
+    yield Cluster()
+    serve.shutdown()
+    ray.shutdown()
+    cluster.shutdown()
+
+
+def test_scale_up(ray_cluster):
+    cluster = ray_cluster
+    head_node = cluster.add_node(num_cpus=3)
+
+    @serve.deployment("D", version="1", num_replicas=1)
+    def D(*args):
+        return os.getpid()
+
+    def get_pids(expected, timeout=30):
+        pids = set()
+        start = time.time()
+        while len(pids) < expected:
+            pids.add(requests.get("http://localhost:8000/D").text)
+            if time.time() - start >= timeout:
+                raise TimeoutError("Timed out waiting for pids.")
+        return pids
+
+    ray.init(head_node.address)
+    serve.start(detached=True)
+    client = serve.connect()
+
+    D.deploy()
+    pids1 = get_pids(1)
+
+    goal_ref = D.options(num_replicas=3).deploy(_blocking=False)
+    assert not client._wait_for_goal(goal_ref, timeout=0.1)
+    assert get_pids(1) == pids1
+
+    # Add a node with another CPU, another replica should get placed.
+    cluster.add_node(num_cpus=1)
+    assert not client._wait_for_goal(goal_ref, timeout=0.1)
+    pids2 = get_pids(2)
+    assert pids1.issubset(pids2)
+
+    # Add a node with another CPU, the final replica should get placed
+    # and the deploy goal should be done.
+    cluster.add_node(num_cpus=1)
+    assert client._wait_for_goal(goal_ref)
+    pids3 = get_pids(3)
+    assert pids2.issubset(pids3)
+
+
+@pytest.mark.skip("Currently hangs due to max_task_retries=-1.")
```

Comment:
@simon-mo this is actually quite bad. Right now the handle doesn't gracefully handle replica failures at all; if a request is enqueued after a replica crashed, the request will hang until the replica is brought back up.

Reply:
I think what we should be doing is not using ray's max_restarts or max_task_retries and instead just restart replicas ourselves

Reply:
Implemented here: https://github.com/ray-project/ray/pull/15047

Reply:
agreed. it's easier to perform app level healthcheck and restart for backends. we should still use ray's fault tolerance mechanism for controller though